src/io.c

```diff
@@ -278,18 +260,18 @@ Char GET_NEXT_CHAR_NO_LC(void)
 
 Char PEEK_NEXT_CHAR(void)
 {
-    assert(IS_CHAR_PUSHBACK_EMPTY());
-
     // store the current character
-    IO()->Pushback = *STATE(In);
+    char c = *STATE(In);
 
     // read next character
     GET_NEXT_CHAR();
 
-    // fake insert the previous character
-    IO()->RealIn = STATE(In);
-    STATE(In) = &IO()->Pushback;
-    return *IO()->RealIn;
+    // insert c back into the line buffer
+    --STATE(In);
```

Comment:
While it should never matter, maybe just add a sanity check GAP_ASSERT(STATE(In) >= IO->Input->line) after this?

Reply:
I've added this and a few more asserts. I also added some comments to explain a bit better what's going on.